File: day15.rb
input = %w[8 13 1 0 18 9]

def part1(input, limit)
  last_number = nil
  memory = {}
  counter = 1

  input.each do |data|
    memory[data] = [counter]
    last_number = data
    counter += 1
  end

  while counter <= limit
    case memory[last_number].length
    when 1
      last_number = '0'
    when 2
      last_number = (memory[last_number][1].to_i - memory[last_number][0].to_i).to_s
    end

    if memory.key? last_number
      memory[last_number] << counter
      memory[last_number].shift if memory[last_number].length > 2
    else
      memory[last_number] = [counter]
    end

    counter += 1
  end

  last_number
end

puts part1(input, 2020)
puts part1(input, 30_000_000)

File: day15.js
const input = ['8', '13', '1', '0', '18', '9'];

function part1(input, limit) {
  let lastNumber = null;
  const memory = {};
  let counter = 1;

  for (const number of input) {
    memory[number] = [counter];
    lastNumber = number;
    counter++;
  }

  do {
    switch(memory[lastNumber].length) {
      case 1:
        lastNumber = '0';
        break;
      case 2:
        lastNumber = (parseInt(memory[lastNumber][1]) - parseInt(memory[lastNumber][0])).toString();
        break;
    }

    if (lastNumber in memory) {
      memory[lastNumber].push(counter);
      if (memory[lastNumber].length > 2) memory[lastNumber].shift();
    } else {
      memory[lastNumber] = [counter];
    }

    counter++;
  } while (counter <= limit);

  console.log(lastNumber);
}

part1(input, 2020);
part1(input, 30000000);
